User Management UI

LKCSmith opened the pull request:
# Purpose
This adds the User Management UI to ops

# Changes
1. Add UI table to list all existing users
2. Ability to search for users on the table (fuzzy search)
3. Pagination for table when 25+
4. Can add a new user
5. Can edit a user's info (username is always uneditable, as an admin, I can edit any user's name, as the user that is currently logged in, I can change my own password)
6. I can delete a user


Note/Asks for reviewers:
1. I would like an extra check on the safety of entering and updating a user's password and make sure this is good from a security POV.

# Checklist

- [ ] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md).
- If docs updated (select one):
  - [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
- [ ] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Closes #357 
Closes #358 
Closes #359 
Closes #360

seanpreston commented on fidesops.toml:
```diff
@@ -25,5 +25,5 @@ DRP_JWT_SECRET="secret"
 TASK_RETRY_DELAY=1
 TASK_RETRY_BACKOFF=1
-REQUIRE_MANUAL_REQUEST_APPROVAL=false
+REQUIRE_MANUAL_REQUEST_APPROVAL=true
```
Can we change this back to `false`? Right now the spec. is for Fidesops to default to automatically approving privacy requests.

seanpreston commented:
Hey @LKCSmith great start! I've spent some time functionally testing this afternoon and found the following to address:

- User is prompted to login inconsistently on view change — were you experiencing this during development?
- No error message in UI for user creation 422s (e.g. password not strong enough) — this should ideally show the same way as if a username or password is omitted, with a different message
- View users permission omitted from list — I suspect this was an oversight in the design? What happens is that new users created are not able to view users in the system, even if they are able to create them
[image:F936A5FB-F44A-46B0-A465-6A13A13BCFE5-32382-000008917E502A12/Screenshot 2022-05-13 at 18.06.30.png]
- Login screen asks for email, should read username for now — not directly related to this PR but would be nice to correct
- Privileges toggles not working correctly (unless this is intentional and I'm not across the decision?)
	- Viewing policies auto toggles viewing datastore connections
	- Create policies auto triggers approve subject requests
	- Can’t remove dependencies ^

seanpreston reviewed (changes requested):
Hey @LKCSmith great start! I've spent some time functionally testing this afternoon and found the following to address:

- User is prompted to login inconsistently on view change — were you experiencing this during development?
- No error message in UI for user creation 422s (e.g. password not strong enough) — this should ideally show the same way as if a username or password is omitted, with a different message
- View users permission omitted from list — I suspect this was an oversight in the design? What happens is that new users created are not able to view users in the system, even if they are able to create them
- Login screen asks for email, should read username for now — not directly related to this PR but would be nice to correct
- Privileges toggles not working correctly (unless this is intentional and I'm not across the decision?)
  - Viewing policies auto toggles viewing datastore connections
  - Create policies auto triggers approve subject requests
  - Can’t remove dependencies ^

LKCSmith commented:
@seanpreston 

`User is prompted to login inconsistently on view change — were you experiencing this during development?`
I wasn't, and I am not experiencing this locally this morning either. Are you noticing a specific navigation around the app that causes this? I've tried jumping around all over but couldn't trigger this bug.

`No error message in UI for user creation 422s (e.g. password not strong enough) — this should ideally show the same way as if a username or password is omitted, with a different message`
Will work on this now 🙂 

`View users permission omitted from list — I suspect this was an oversight in the design? What happens is that new users created are not able to view users in the system, even if they are able to create them`
Added, although I don't think that alone resolves the issue. I also added a checkbox for 'View users'. Should this as well as 'View roles' be automatically checked for each user created?

`Login screen asks for email, should read username for now — not directly related to this PR but would be nice to correct`
Fixed!

```
Privileges toggles not working correctly (unless this is intentional and I'm not across the decision?)
Viewing policies auto toggles viewing datastore connections
Create policies auto triggers approve subject requests
Can’t remove dependencies ^
```
I'm not experiencing these bugs locally either. Maybe you can demonstrate to me on a call?

allisonking commented on clients/admin-ui/src/features/user-management/EditUserForm.tsx:
```diff
@@ -0,0 +1,314 @@
+            old_password: existingUser?.password,
+            new_password: values.password,
+          };
+          console.log(updatePasswordbody);
```
and while you're at it.... another nit: `updatePasswordbody` --> `updatePasswordBody`

allisonking commented on clients/admin-ui/src/features/user-management/EditUserForm.tsx:
```diff
@@ -0,0 +1,314 @@
+                    id={`scopes-${policy.privilege}`}
+                    name="scopes"
+                    value={policy.scope}
+                    isDisabled={policy.scope === 'privacy-request:read'}
```
out of curiosity, how did you get rid of that uncontrolled vs controlled input console log msg?

allisonking commented on clients/admin-ui/src/features/user/types.ts:
```diff
@@ -0,0 +1,94 @@
+export const userPrivilegesArray: UserPrivileges[] = [
+  {
+    privilege: 'View subject requests',
+    scope: 'privacy-request:read',
```
I think one day, not necessarily in this PR, we will want to pull these out as constants, such as like what is in the backend https://github.com/ethyca/fidesops/blob/user-management/src/fidesops/api/v1/scope_registry.py so that it'll be a lot easier in code to refer to the permissions

allisonking reviewed (commented):
Wow, nice job getting this all set up! 💪  

Here are also some comments from playing around locally:

* Creating a new user doesn't tell me which fields are required but doesn't let me save if there's no password
  * Only says it's required if you have touched it
* Unchecking a permission and saving seems to work, but get a POST error in the console
```
POST http://0.0.0.0:8080/api/v1/api/v1/user/fid_3abb8caf-29a5-48d5-a4d9-b8c362ccf9c6/reset-password 404 (Not Found)
```
* No error when trying to delete a user and the usernames don't match
  * Can't tab between the two fields of username and enter username again
* Can edit the superuser and make it so they are stuck and can't edit anymore :( 
  * Repro steps: 
    1. Edit your superuser and remove some permissions (I removed "Approve subject requests", "view datastore connections", "create or update datastore connections", "delete datastore connections")
    2. After you save, there will be no users listed in the User Management page (not sure if one of those perms I removed controls that? I purposefully didn't remove the "view users" policy)
  * Looks like this is because of a 403 forbidden, though I'm not sure which of those permissions makes me unable to view users anymore. Also, it seems undesirable to let the superuser get themselves into this state.

elliotbonneville commented:
Absolutely crushing it, Laura. Love how fast you onboarded and that you're knocking out such a big feature already. Great work! 🎉 

I think the functional testing Sean and Allison did is quite thorough. In addition to what they've found, I've noticed the following things:

- Logging in, then navigating around User Management pages triggers logout sporadically (and then you need to log in twice)
- Clicking links in the breadcrumbs nav triggers a full page reload (should use Next link component to avoid that)
- Spacing issues throughout the layout (mostly just padding tweaks for eyecandy – not critical path)
- Submit button for creating a user should read 'Create' instead of 'Save'
- There shouldn't be a placeholder in the login screen
- Can't tab through inputs in the Delete User modal for some reason
- Password input for new users is a bit laggy (this may indicate state mismanagement on some level)
- Username is not passed to dropdown menu

Of these, the only critical bug is the sporadic logout issues. However, since the functionality is there, I'm comfortable merging, provided we address this bug this week (something I'm happy to take on today / tomorrow). The rest of these don't need to be addressed in these PR either – I'm just dropping them here for future reference.

I also have a number of stylistic / cleanup related nitpicks w/r/t the code. I began to make individual comments to address them, but I think in this case it might be more efficient for me to clean them up on my own after this PR is merged, and then we can do a bit of a post mortem next week and review some of the changes I made.

And to be clear, when doing PR reviews in general, it's generally my opinion that any syntax or style things that come up are more a function of the scope of work and the time constraint around it than of somebody's personal coding style. So don't feel like any changes I make are because mistakes were made or that there were glaring oversights. I totally recognize that it's more just the sheer volume of code leaves plenty of spaces for bugs to hide and makes cleanup difficult + time-consuming :)

Finally, I'd like to address the scope of this PR – I think going forward we'll want to try to break features down into smaller, more manageable PRs so that we don't get into a place of overwhelm like this again. That would help with both finding bugs, general feelings of overwhelm, and also give more time and space to do cleanup.

Again, amazing job with this! Very happy to see it land. 🥳
